File: llm-service/infrastructure/event_bus.py
"""
Enterprise Event Bus Implementation for LLM Service
Provides reliable, scalable event-driven communication using Kafka
"""

import json
import asyncio
import logging
from typing import Dict, Any, Optional, Callable, List
from datetime import datetime
from uuid import uuid4
from dataclasses import dataclass, asdict
from kafka import KafkaProducer, KafkaConsumer
from kafka.errors import KafkaError
import structlog

logger = structlog.get_logger(__name__)

@dataclass
class BaseEvent:
    id: str
    type: str
    timestamp: datetime
    correlation_id: str
    source: str
    version: str

@dataclass
class LLMEvent(BaseEvent):
    user_id: int
    model: str
    operation: str
    data: Dict[str, Any]
    metadata: Optional[Dict[str, Any]] = None

@dataclass
class ContentEvent(BaseEvent):
    user_id: int
    content_type: str
    data: Dict[str, Any]
    metadata: Optional[Dict[str, Any]] = None

@dataclass
class SystemEvent(BaseEvent):
    service: str
    data: Dict[str, Any]
    severity: str = 'medium'

class LLMEventBus:
    """Enterprise Event Bus for LLM Service"""
    
    TOPICS = {
        'LLM_EVENTS': 'llm-events',
        'CONTENT_EVENTS': 'content-events',
        'SYSTEM_EVENTS': 'system-events',
        'ERROR_EVENTS': 'error-events'
    }
    
    def __init__(self, kafka_config: Optional[Dict[str, Any]] = None):
        self.kafka_config = kafka_config or {
            'bootstrap_servers': ['kafka:29092'],
            'client_id': 'llm-service',
            'value_serializer': lambda v: json.dumps(v, default=str).encode('utf-8'),
            'value_deserializer': lambda m: json.loads(m.decode('utf-8')),
            'acks': 'all',
            'retries': 5,
            'retry_backoff_ms': 1000,
            'max_in_flight_requests_per_connection': 1,
            'enable_idempotence': True
        }
        
        self.producer: Optional[KafkaProducer] = None
        self.consumers: Dict[str, KafkaConsumer] = {}
        self.is_connected = False
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 10
        
    async def initialize(self):
        """Initialize the event bus"""
        try:
            logger.info("Initializing LLM Event Bus...")
            
            # Initialize producer
            self.producer = KafkaProducer(**self.kafka_config)
            
            # Create topics if they don't exist
            await self._create_topics()
            
            self.is_connected = True
            self.reconnect_attempts = 0
            
            logger.info("LLM Event Bus initialized successfully")
            
        except Exception as error:
            logger.error("Failed to initialize Event Bus", error=str(error))
            await self._handle_reconnection()
            raise error
    
    async def publish(self, event: BaseEvent):
        """Publish an event to the appropriate topic"""
        if not self.is_connected or not self.producer:
            raise Exception("Event Bus not connected")
        
        try:
            topic = self._get_topic_for_event_type(event.type)
            message = asdict(event)
            
            # Convert datetime to string for JSON serialization
            message['timestamp'] = event.timestamp.isoformat()
            
            future = self.producer.send(
                topic,
                key=str(event.user_id if hasattr(event, 'user_id') else event.id),
                value=message,
                headers={
                    'event-type': event.type.encode('utf-8'),
                    'correlation-id': event.correlation_id.encode('utf-8'),
                    'source': event.source.encode('utf-8'),
                    'version': event.version.encode('utf-8')
                }
            )
            
            # Wait for the message to be sent
            record_metadata = future.get(timeout=10)
            
            logger.debug(
                "Event published successfully",
                event_id=event.id,
                event_type=event.type,
                topic=topic,
                partition=record_metadata.partition,
                offset=record_metadata.offset,
                correlation_id=event.correlation_id
            )
            
        except Exception as error:
            logger.error(
                "Failed to publish event",
                error=str(error),
                event_id=event.id,
                event_type=event.type
            )
            
            # Publish error event
            await self.publish_system_event(
                'system.error',
                'llm-service',
                {
                    'error': str(error),
                    'event_id': event.id,
                    'event_type': event.type
                },
                'high'
            )
            
            raise error
    
    async def subscribe(
        self,
        event_type: str,
        handler: Callable[[Dict[str, Any]], None],
        group_id: Optional[str] = None
    ):
        """Subscribe to events of a specific type"""
        group_id = group_id or f"llm-{event_type}-consumer"
        
        try:
            topic = self._get_topic_for_event_type(event_type)
            
            consumer_config = {
                **self.kafka_config,
                'group_id': group_id,
                'auto_offset_reset': 'latest',
                'enable_auto_commit': True,
                'session_timeout_ms': 30000,
                'heartbeat_interval_ms': 3000
            }
            
            # Remove producer-specific configs
            consumer_config.pop('acks', None)
            consumer_config.pop('retries', None)
            consumer_config.pop('enable_idempotence', None)
            consumer_config.pop('max_in_flight_requests_per_connection', None)
            
            consumer = KafkaConsumer(topic, **consumer_config)
            self.consumers[f"{group_id}-{event_type}"] = consumer
            
            # Start consuming in background
            asyncio.create_task(self._consume_messages(consumer, handler, event_type))
            
            logger.info(
                "Subscribed to events",
                event_type=event_type,
                topic=topic,
                group_id=group_id
            )
            
        except Exception as error:
            logger.error(
                "Failed to subscribe to events",
                error=str(error),
                event_type=event_type,
                group_id=group_id
            )
            raise error
    
    async def publish_llm_event(
        self,
        event_type: str,
        user_id: int,
        model: str,
        operation: str,
        data: Dict[str, Any],
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Publish an LLM-specific event"""
        event = LLMEvent(
            id=str(uuid4()),
            type=event_type,
            user_id=user_id,
            model=model,
            operation=operation,
            data=data,
            metadata=metadata,
            timestamp=datetime.utcnow(),
            correlation_id=str(uuid4()),
            source='llm-service',
            version='1.0.0'
        )
        
        await self.publish(event)
    
    async def publish_content_event(
        self,
        event_type: str,
        user_id: int,
        content_type: str,
        data: Dict[str, Any],
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Publish a content-specific event"""
        event = ContentEvent(
            id=str(uuid4()),
            type=event_type,
            user_id=user_id,
            content_type=content_type,
            data=data,
            metadata=metadata,
            timestamp=datetime.utcnow(),
            correlation_id=str(uuid4()),
            source='llm-service',
            version='1.0.0'
        )
        
        await self.publish(event)
    
    async def publish_system_event(
        self,
        event_type: str,
        service: str,
        data: Dict[str, Any],
        severity: str = 'medium'
    ):
        """Publish a system event"""
        event = SystemEvent(
            id=str(uuid4()),
            type=event_type,
            service=service,
            data=data,
            severity=severity,
            timestamp=datetime.utcnow(),
            correlation_id=str(uuid4()),
            source='llm-service',
            version='1.0.0'
        )
        
        await self.publish(event)
    
    def get_health_status(self) -> Dict[str, Any]:
        """Get health status of the event bus"""
        return {
            'connected': self.is_connected,
            'reconnect_attempts': self.reconnect_attempts,
            'active_consumers': len(self.consumers),
            'topics': list(self.TOPICS.values())
        }
    
    async def shutdown(self):
        """Gracefully shutdown the event bus"""
        logger.info("Shutting down LLM Event Bus...")
        
        try:
            # Close all consumers
            for key, consumer in self.consumers.items():
                consumer.close()
                logger.debug(f"Consumer {key} closed")
            
            self.consumers.clear()
            
            # Close producer
            if self.producer:
                self.producer.close()
            
            self.is_connected = False
            logger.info("LLM Event Bus shutdown completed")
            
        except Exception as error:
            logger.error("Error during Event Bus shutdown", error=str(error))
            raise error
    
    async def _create_topics(self):
        """Create required topics"""
        # This would typically use Kafka Admin API
        # For now, topics are created automatically by Kafka
        logger.info("Kafka topics will be created automatically")
    
    def _get_topic_for_event_type(self, event_type: str) -> str:
        """Get appropriate topic for event type"""
        if event_type.startswith('llm.'):
            return self.TOPICS['LLM_EVENTS']
        elif event_type.startswith('content.'):
            return self.TOPICS['CONTENT_EVENTS']
        elif event_type.startswith('system.'):
            return self.TOPICS['SYSTEM_EVENTS']
        else:
            return self.TOPICS['ERROR_EVENTS']
    
    async def _consume_messages(
        self,
        consumer: KafkaConsumer,
        handler: Callable[[Dict[str, Any]], None],
        event_type: str
    ):
        """Consume messages from Kafka"""
        try:
            for message in consumer:
                try:
                    event_data = message.value
                    correlation_id = event_data.get('correlation_id', 'unknown')
                    
                    logger.debug(
                        "Processing event",
                        event_id=event_data.get('id'),
                        event_type=event_data.get('type'),
                        partition=message.partition,
                        offset=message.offset,
                        correlation_id=correlation_id
                    )
                    
                    # Execute handler with timeout
                    await asyncio.wait_for(
                        asyncio.create_task(handler(event_data)),
                        timeout=30.0
                    )
                    
                    logger.debug(
                        "Event processed successfully",
                        event_id=event_data.get('id'),
                        event_type=event_data.get('type'),
                        correlation_id=correlation_id
                    )
                    
                except asyncio.TimeoutError:
                    logger.error(
                        "Event handler timeout",
                        event_type=event_type,
                        partition=message.partition,
                        offset=message.offset
                    )
                except Exception as error:
                    logger.error(
                        "Error processing event",
                        error=str(error),
                        event_type=event_type,
                        partition=message.partition,
                        offset=message.offset
                    )
                    
        except Exception as error:
            logger.error(
                "Error in message consumer",
                error=str(error),
                event_type=event_type
            )
    
    async def _handle_reconnection(self):
        """Handle reconnection logic"""
        if self.reconnect_attempts >= self.max_reconnect_attempts:
            logger.error("Max reconnection attempts reached")
            return
        
        self.reconnect_attempts += 1
        logger.info(f"Attempting to reconnect ({self.reconnect_attempts}/{self.max_reconnect_attempts})...")
        
        await asyncio.sleep(5)  # Wait 5 seconds before retry
        
        try:
            await self.initialize()
        except Exception as error:
            logger.error("Reconnection failed", error=str(error))
            await self._handle_reconnection()

# Singleton instance
event_bus = LLMEventBus()

File: telegram-bot/src/infrastructure/eventBus.ts
/**
 * Enterprise Event Bus Implementation
 * Provides reliable, scalable event-driven communication using Kafka
 */

import { Kafka, Producer, Consumer, EachMessagePayload, KafkaConfig } from 'kafkajs';
import { logger } from '../utils/logger';
import { v4 as uuidv4 } from 'uuid';
import { EventEmitter } from 'events';

// Event Types
export interface BaseEvent {
  id: string;
  type: string;
  timestamp: Date;
  correlationId: string;
  source: string;
  version: string;
}

export interface TelegramEvent extends BaseEvent {
  type: 'telegram.message' | 'telegram.callback' | 'telegram.command' | 'telegram.error' | 'telegram.webhook';
  userId: number;
  data: any;
  metadata?: Record<string, any> | undefined;
}

export interface UserEvent extends BaseEvent {
  type: 'user.created' | 'user.updated' | 'user.deleted' | 'user.activity';
  userId: number;
  action: string;
  metadata: Record<string, any>;
}

export interface ContentEvent extends BaseEvent {
  type: 'content.generated' | 'content.failed' | 'content.cached';
  userId: number;
  contentType: string;
  data: any;
  metadata?: Record<string, any> | undefined;
}

export interface SystemEvent extends BaseEvent {
  type: 'system.health' | 'system.error' | 'system.metric' | 'system.campaign';
  service: string;
  data: any;
  severity?: 'low' | 'medium' | 'high' | 'critical';
}

export type Event = TelegramEvent | UserEvent | ContentEvent | SystemEvent;

// Event Handler Type
export type EventHandler<T extends Event = Event> = (event: T) => Promise<void>;

// Topic Configuration
export const TOPICS = {
  TELEGRAM_EVENTS: 'telegram-events',
  USER_EVENTS: 'user-events',
  CONTENT_EVENTS: 'content-events',
  ANALYTICS_EVENTS: 'analytics-events',
  SYSTEM_EVENTS: 'system-events',
  ERROR_EVENTS: 'error-events'
} as const;

export class EnterpriseEventBus extends EventEmitter {
  private kafka: Kafka;
  private producer: Producer;
  private consumers: Map<string, Consumer> = new Map();
  private isConnected: boolean = false;
  private reconnectAttempts: number = 0;
  private maxReconnectAttempts: number = 10;
  private reconnectDelay: number = 5000;

  constructor(config?: Partial<KafkaConfig>) {
    super();

    // Parse Kafka brokers from environment
    const kafkaBrokers = process.env.KAFKA_BROKERS?.split(',').map(b => b.trim()) || ['localhost:9092'];
    logger.info('Kafka brokers configuration:', { brokers: kafkaBrokers, env: process.env.KAFKA_BROKERS });

    const defaultConfig: KafkaConfig = {
      clientId: 'telegram-bot-service',
      brokers: kafkaBrokers,
      connectionTimeout: 30000,
      requestTimeout: 30000,
      retry: {
        initialRetryTime: 100,
        retries: 5, // Limit retries to prevent infinite loops
        maxRetryTime: 30000,
        factor: 2,
        multiplier: 2,
        restartOnFailure: async (e: Error) => {
          logger.error('Kafka restart on failure:', e);
          return false; // Don't restart automatically to prevent crash loops
        }
      },
      logLevel: 4, // ERROR level to suppress warnings
    };

    this.kafka = new Kafka({ ...defaultConfig, ...config });
    // Use recommended idempotent producer configuration
    this.producer = this.kafka.producer({
      maxInFlightRequests: 5,
      idempotent: true,
      retry: {
        retries: Number.MAX_SAFE_INTEGER, // Unlimited retries for idempotent producer
        initialRetryTime: 300,
        maxRetryTime: 10000
      }
    });

    this.setupErrorHandlers();
  }

  /**
   * Initialize the event bus
   */
  async initialize(): Promise<void> {
    try {
      logger.info('Initializing Enterprise Event Bus...');

      // Add timeout to prevent hanging
      const connectPromise = this.producer.connect();
      const timeoutPromise = new Promise((_, reject) =>
        setTimeout(() => reject(new Error('Kafka connection timeout')), 10000)
      );

      await Promise.race([connectPromise, timeoutPromise]);
      await this.createTopics();

      this.isConnected = true;
      this.reconnectAttempts = 0;

      logger.info('Enterprise Event Bus initialized successfully');
      this.emit('connected');

    } catch (error) {
      logger.error('Failed to initialize Event Bus:', error);
      // Don't throw error to prevent container crash
      logger.warn('Continuing without Kafka event bus');
      this.isConnected = false;
    }
  }

  /**
   * Publish an event to the appropriate topic
   */
  async publish<T extends Event>(event: T): Promise<void> {
    if (!this.isConnected) {
      // Gracefully handle disconnected state instead of throwing
      logger.warn(`Event Bus not connected, skipping publish for event type: ${event.type}`);
      return;
    }

    try {
      const topic = this.getTopicForEventType(event.type);
      const message = {
        key: (event as any).userId?.toString() || event.id,
        value: JSON.stringify(event),
        timestamp: event.timestamp.getTime().toString(),
        headers: {
          'event-type': event.type,
          'correlation-id': event.correlationId,
          'source': event.source,
          'version': event.version
        }
      };

      await this.producer.send({
        topic,
        messages: [message]
      });

      logger.debug('Event published successfully', {
        eventId: event.id,
        eventType: event.type,
        topic,
        correlationId: event.correlationId
      });

      // Emit local event for immediate handlers
      this.emit(event.type, event);

    } catch (error) {
      logger.error('Failed to publish event:', error, {
        eventId: event.id,
        eventType: event.type
      });
      
      // Emit error event
      await this.publishSystemEvent('system.error', 'event-bus', {
        error: error instanceof Error ? error.message : 'Unknown error',
        eventId: event.id,
        eventType: event.type
      }, 'high');
      
      throw error;
    }
  }

  /**
   * Subscribe to events of a specific type
   */
  async subscribe<T extends Event>(
    eventType: string,
    handler: EventHandler<T>,
    options: {
      groupId?: string;
      fromBeginning?: boolean;
      autoCommit?: boolean;
    } = {}
  ): Promise<void> {
    const {
      groupId = `telegram-bot-${eventType}-consumer`,
      fromBeginning = false,
      autoCommit = true
    } = options;

    try {
      const topic = this.getTopicForEventType(eventType);
      const consumer = this.kafka.consumer({
        groupId,
        sessionTimeout: 30000,
        heartbeatInterval: 3000,
        maxWaitTimeInMs: 5000,
        retry: {
          retries: 5,
          initialRetryTime: 300,
          maxRetryTime: 30000
        }
      });

      await consumer.connect();
      await consumer.subscribe({ topic, fromBeginning });

      await consumer.run({
        autoCommit,
        eachMessage: async ({ topic, partition, message }: EachMessagePayload) => {
          try {
            if (!message.value) return;

            const event = JSON.parse(message.value.toString()) as T;
            
            // Add tracing information
            const correlationId = message.headers?.['correlation-id']?.toString() || event.correlationId;
            
            logger.debug('Processing event', {
              eventId: event.id,
              eventType: event.type,
              topic,
              partition,
              offset: message.offset,
              correlationId
            });

            // Execute handler with timeout
            await Promise.race([
              handler(event),
              new Promise((_, reject) => 
                setTimeout(() => reject(new Error('Handler timeout')), 30000)
              )
            ]);

            logger.debug('Event processed successfully', {
              eventId: event.id,
              eventType: event.type,
              correlationId
            });

          } catch (error) {
            logger.error('Error processing event:', error, {
              topic,
              partition,
              offset: message.offset,
              eventType
            });

            // Publish error event
            await this.publishSystemEvent('system.error', 'event-consumer', {
              error: error instanceof Error ? error.message : 'Unknown error',
              topic,
              partition,
              offset: message.offset,
              eventType
            }, 'high');

            // Don't throw - let Kafka handle retry logic
          }
        }
      });

      this.consumers.set(`${groupId}-${eventType}`, consumer);
      
      logger.info('Subscribed to events', {
        eventType,
        topic,
        groupId
      });

    } catch (error) {
      logger.error('Failed to subscribe to events:', error, {
        eventType,
        groupId
      });
      throw error;
    }
  }

  /**
   * Publish a Telegram event
   */
  async publishTelegramEvent(
    type: TelegramEvent['type'],
    userId: number,
    data: any,
    metadata?: Record<string, any>
  ): Promise<void> {
    const event: TelegramEvent = {
      id: uuidv4(),
      type,
      userId,
      data,
      metadata,
      timestamp: new Date(),
      correlationId: uuidv4(),
      source: 'telegram-bot',
      version: '1.0.0'
    };

    await this.publish(event);
  }

  /**
   * Publish a user event
   */
  async publishUserEvent(
    type: UserEvent['type'],
    userId: number,
    action: string,
    metadata: Record<string, any>
  ): Promise<void> {
    const event: UserEvent = {
      id: uuidv4(),
      type,
      userId,
      action,
      metadata,
      timestamp: new Date(),
      correlationId: uuidv4(),
      source: 'telegram-bot',
      version: '1.0.0'
    };

    await this.publish(event);
  }

  /**
   * Publish a content event
   */
  async publishContentEvent(
    type: ContentEvent['type'],
    userId: number,
    contentType: string,
    data: any,
    metadata?: Record<string, any>
  ): Promise<void> {
    const event: ContentEvent = {
      id: uuidv4(),
      type,
      userId,
      contentType,
      data,
      metadata,
      timestamp: new Date(),
      correlationId: uuidv4(),
      source: 'telegram-bot',
      version: '1.0.0'
    };

    await this.publish(event);
  }

  /**
   * Publish an LLM event
   */
  async publishLLMEvent(
    type: string,
    userId: number,
    model: string,
    operation: string,
    data: any,
    metadata?: Record<string, any> | undefined
  ): Promise<void> {
    const event: ContentEvent = {
      id: uuidv4(),
      type: 'content.generated',
      userId,
      contentType: model,
      data: {
        model,
        operation,
        ...data
      },
      metadata: metadata || undefined,
      timestamp: new Date(),
      correlationId: uuidv4(),
      source: 'telegram-bot',
      version: '1.0.0'
    };

    await this.publish(event);
  }

  /**
   * Publish a system event
   */
  async publishSystemEvent(
    type: SystemEvent['type'],
    service: string,
    data: any,
    severity: SystemEvent['severity'] = 'medium'
  ): Promise<void> {
    const event: SystemEvent = {
      id: uuidv4(),
      type,
      service,
      data,
      severity,
      timestamp: new Date(),
      correlationId: uuidv4(),
      source: 'telegram-bot',
      version: '1.0.0'
    };

    await this.publish(event);
  }

  /**
   * Get health status of the event bus
   */
  getHealthStatus(): {
    connected: boolean;
    reconnectAttempts: number;
    activeConsumers: number;
    topics: string[];
  } {
    return {
      connected: this.isConnected,
      reconnectAttempts: this.reconnectAttempts,
      activeConsumers: this.consumers.size,
      topics: Object.values(TOPICS)
    };
  }

  /**
   * Gracefully shutdown the event bus
   */
  async shutdown(): Promise<void> {
    logger.info('Shutting down Enterprise Event Bus...');

    try {
      // Disconnect all consumers
      for (const [key, consumer] of this.consumers) {
        await consumer.disconnect();
        logger.debug(`Consumer ${key} disconnected`);
      }
      this.consumers.clear();

      // Disconnect producer
      await this.producer.disconnect();
      
      this.isConnected = false;
      logger.info('Enterprise Event Bus shutdown completed');
      
    } catch (error) {
      logger.error('Error during Event Bus shutdown:', error);
      throw error;
    }
  }

  /**
   * Create required topics
   */
  private async createTopics(): Promise<void> {
    const admin = this.kafka.admin();
    
    try {
      await admin.connect();
      
      const topics = Object.values(TOPICS).map(topic => ({
        topic,
        numPartitions: 3,
        replicationFactor: 1,
        configEntries: [
          { name: 'cleanup.policy', value: 'delete' },
          { name: 'retention.ms', value: '604800000' }, // 7 days
          { name: 'compression.type', value: 'gzip' },
          { name: 'min.insync.replicas', value: '1' }
        ]
      }));

      await admin.createTopics({ topics });
      logger.info('Kafka topics created/verified');
      
    } catch (error) {
      logger.warn('Error creating topics (may already exist):', error);
    } finally {
      await admin.disconnect();
    }
  }

  /**
   * Get appropriate topic for event type
   */
  private getTopicForEventType(eventType: string): string {
    if (eventType.startsWith('telegram.')) return TOPICS.TELEGRAM_EVENTS;
    if (eventType.startsWith('user.')) return TOPICS.USER_EVENTS;
    if (eventType.startsWith('content.')) return TOPICS.CONTENT_EVENTS;
    if (eventType.startsWith('analytics.')) return TOPICS.ANALYTICS_EVENTS;
    if (eventType.startsWith('system.')) return TOPICS.SYSTEM_EVENTS;
    return TOPICS.ERROR_EVENTS;
  }

  /**
   * Setup error handlers
   */
  private setupErrorHandlers(): void {
    this.producer.on('producer.disconnect', () => {
      logger.warn('Producer disconnected');
      this.isConnected = false;
    });

    this.producer.on('producer.network.request_timeout', (payload: any) => {
      logger.warn('Producer network timeout:', payload);
    });
  }

  /**
   * Handle reconnection logic
   */
  private async handleReconnection(): Promise<void> {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      logger.error('Max reconnection attempts reached');
      this.emit('error', new Error('Max reconnection attempts reached'));
      return;
    }

    this.reconnectAttempts++;
    logger.info(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts})...`);

    await new Promise(resolve => setTimeout(resolve, this.reconnectDelay));

    try {
      await this.initialize();
    } catch (error) {
      logger.error('Reconnection failed:', error);
      await this.handleReconnection();
    }
  }
}

// Singleton instance
export const eventBus = new EnterpriseEventBus();
